bee/src/main/java/com/apisense/bee/ui/adapter/CropFilter.java

```diff
@@ -0,0 +1,58 @@
+package com.apisense.bee.ui.adapter;
+
+import android.widget.Filter;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import io.apisense.sdk.core.store.Crop;
+
+class CropFilter extends Filter {
+    private final AvailableExperimentsRecyclerAdapter adapter;
+    private final List<Crop> crops;
+
+    CropFilter(AvailableExperimentsRecyclerAdapter adapter, List<Crop> crops) {
+        this.adapter = adapter;
+        this.crops = crops;
+    }
+
+    @Override
+    protected FilterResults performFiltering(CharSequence constraint) {
+        if (constraint.length() == 0) {
+            return buildResult(crops);
+        } else {
+            return searchSubstring(constraint);
+        }
+    }
+
+    private FilterResults searchSubstring(CharSequence constraint) {
+        ArrayList<Crop> filtered = new ArrayList<>();
+        for (Crop crop : crops) {
+            String filter = constraint.toString();
+            if (isContained(crop.getName(), filter)
+                    || isContained(crop.getOwner(), filter)
+                    || isContained(crop.getLongDescription(), filter)
+                    || isContained(crop.getShortDescription(), filter)) {
+                filtered.add(crop);
+            }
+        }
+        return buildResult(filtered);
+    }
+
+    private boolean isContained(String field, String constraint) {
+        return field.toLowerCase().contains(constraint.toLowerCase());
+    }
+
+    private FilterResults buildResult(List<Crop> filtered) {
+        FilterResults results = new FilterResults();
+        results.values = filtered;
+        results.count = filtered.size();
+        return results;
+    }
+
+    @Override
+    protected void publishResults(CharSequence constraint, FilterResults results) {
+        adapter.setAvailableCrops((List<Crop>) results.values);
```

Comment:
`Unchecked cast: 'java.lang.Object' to 'java.util.List<io.apisense.sdk.core.store.Crop>'`

Reply:
We are filling the `result` object in the same class, in the method `buildResult` thus the cast is in fact safe.

There are two possibilities here:
- Consider that the cast is safe because we handle everything in the same class
- Wrap a `List<Crop>` in a specific result class in order to only cast `result.values` into this class and avoid the unchecked cast.

Reply:
The second one seems to be better in my point of view :).